### ISSUE TITLE
Dat API & avoiding host-device transfers in OpenCL

### ISSUE THREAD

**Issue by gsigms** [OP2/PyOP2#48]
Laurence and I have been having a discussion about the Dat.data accessor and OpenCL device-host transfer recently, here is a summary:

Following Laurence comment in https://github.com/OP2/PyOP2/pull/44:

> Ideally I think _dirty should not be a bool, but an enum, describing:
> 
> clean (host and device agree on data)
> only exists on host (not yet uploaded to device)
> modified on device (needs fetching to host)
> modified on host (needs copying to device)
> 
> The last option is because the .data property acts a bit like op_fetch_data in the C api. Before the first par_loop call, the user > owns the data. After a par_loop call, OP2 owns the data /until/ someone does x = foo.data. Now the data might be modified by the user. So OP2 needs to re-upload it before the next par_loop that uses it. This is not the case is x = foo.data is not called.

We discussed 3 ways to achieve this:

A. Use explicit Dat setter and getter.
B. "Wrap" every numpy array accessor to track modification of the values.
C. Decouple the API: provide a data and data_readonly accessors: whenever one access Dat.data, the data get refreshed on the device before executing the next par_loop.

From our discussion:
### Solution A.

It is far from elegant, plus to ensure consistency between backends, the Dat getter would have to return a copy of the values not a reference. This would forbid code like:

``` python
x.data[0] = 0
```

instead one would have to use:

``` python
d = x.data
d[0] = 0
x.data = d
```
### Solution B.

It looks more elegant, but there are concerns about performance and applicability in the case of Fluidity.
### Solution C.

It would not have the explicit setters while still allowing to update device memory only when necessary in the case of opencl. But the behavior could become tricky to understand, for example:

``` python
foo = x.data
par_loop(..., x(..., RW), ...) # x.copy_host_to_device()
foo[0] = 0
par_loop(..., x(..., RW), ...) # do not update
```

``` python
foo = x.data
par_loop(..., x(..., RW), ...) # x.copy_host_to_device()
foo[0] = 0 # silent
bar = x.data # overwrite the previous statement with device content
foo[1] = 1 # seen
bar[2] = 2 # seen
par_loop(..., x(..., RW), ...) # x.copy_host_to_device() (because of bar)
```

This is kinda tricky, moreover this would require "fixing" the sequential backend to make it adopt an ugly behavior!

Bottom line, solution A. seems the least worse solution to me atm. Does anyone see a better solution ?
